sdk/go/src/sawtooth_sdk/processor/state.go

```diff
@@ -202,3 +204,110 @@ func (self *Context) Get(addresses []string) (map[string][]byte, error) {
 func (self *Context) Set(pairs map[string][]byte) ([]string, error) {
 	return self.SetState(pairs)
 }
+
+func (self *Context) AddReceiptData(data_type string, data []byte) error {
+	// Append the data to the transaction receipt and set the type
+	request := &processor_pb2.TpAddReceiptDataRequest{
+		ContextId: self.contextId,
+		DataType:  data_type,
+		Data:      data,
+	}
+	bytes, err := proto.Marshal(request)
+	if err != nil {
+		return fmt.Errorf("Failed to marshal: %v", err)
+	}
+
+	// Send the message and get the response
+	corrId, err := self.connection.SendNewMsg(
+		validator_pb2.Message_TP_ADD_RECEIPT_DATA_REQUEST, bytes,
+	)
+	if err != nil {
+		return fmt.Errorf("Failed to add receipt data: %v", err)
+	}
+
+	_, msg, err := self.connection.RecvMsgWithId(corrId)
+	if err != nil {
+		return fmt.Errorf("Failed to receive TpAddReciptDataResponse: %v", err)
+	}
+	if msg.GetCorrelationId() != corrId {
+		return fmt.Errorf(
+			"Expected message with correlation id %v but got %v",
+			corrId, msg.GetCorrelationId(),
+		)
+	}
+
+	if msg.GetMessageType() != validator_pb2.Message_TP_ADD_RECEIPT_DATA_RESPONSE {
+		return fmt.Errorf(
+			"Expected TP_ADD_RECEIPT_DATA_RESPONSE but got %v", msg.GetMessageType(),
+		)
+	}
+
+	// Parse the result
+	response := &processor_pb2.TpAddReceiptDataResponse{}
+	err = proto.Unmarshal(msg.Content, response)
+	if err != nil {
+		return fmt.Errorf("Failed to unmarshal TpAddReceiptDataResponse: %v", err)
+	}
+
+	// Use a switch in case new Status values are added
+	switch response.GetStatus() {
+	case processor_pb2.TpAddReceiptDataResponse_ERROR:
+		return fmt.Errorf("Failed to add receipt data")
+	}
+
+	return nil
+}
+
+func (self *Context) AddEvent(event_type string, attributes []*Event.Event_Attribute, event_data []byte) error {
```

Comment:
Either have `AddEvent` take an `Event` as an argument, or replace `Event.Event_Attribute` with an SDK defined pass-by-value struct and construct the whole event within the function. This makes it less awkward to use the `AddEvent` method because you can omit the type when constructing a literal. The function header would look like: `func (self *Context) AddEvent(event_type string, attributes []Attribute, event_data []byte) error` then. See https://play.golang.org/p/6Nv6KwAMce